### ISSUE TITLE
Preview link not correctly set when opening the backend record page (act=edit) directly after login

### ISSUE THREAD

**Issue by Defcon0** [contao/contao#2580]
**Affected version(s)**

4.4, 4.9 (probably 4.10 also)

**Description**

After opening e.g. an event edit page in contao backend in the usual way, i.e. by clicking "Events" and going into the archive and then in the edit page of the event, the preview link is correctly set (`BackendMain` -> `frontendFile` property).

Now some of the our customers have very short session times leading to the record page being reloaded, logged out having to log in again. After this login the `CURRENT_ID` is not set so the following if is false:

```
// vendor/contao/core-bundle/src/Resources/contao/controllers/BackendMain.php

// Front end preview links
if (\defined('CURRENT_ID') && CURRENT_ID != '')
{
	if (\Input::get('do') == 'page')
	{
		$this->Template->frontendFile = '?page=' . CURRENT_ID;
	}
	elseif (\Input::get('do') == 'article' && ($objArticle = \ArticleModel::findByPk(CURRENT_ID)) !== null)
	{
		$this->Template->frontendFile = '?page=' . $objArticle->pid;
	}
	elseif (\Input::get('do') != '')
	{
		$event = new PreviewUrlCreateEvent(\Input::get('do'), CURRENT_ID);
		\System::getContainer()->get('event_dispatcher')->dispatch(ContaoCoreEvents::PREVIEW_URL_CREATE, $event);

		if (($strQuery = $event->getQuery()) !== null)
		{
			$this->Template->frontendFile = '?' . $strQuery;
		}
	}
}
```

and the `frontendFile` is not being set.

Currently for Contao 4.4 I built a rather ugly workaround using the parseTemplate hook ;-) But I don't quite know what a good solution could be. Do you have ideas?

Thanks!

**Comment by leofeyer** [comment]
Does the issue really occur in Contao 4.9? Because there is no `CURRENT_ID` in the class anymore.

**Comment by Defcon0** [comment]
Could be, but if you do the steps above, you can see the wrong behavior in contao 4.9 also.

Here the workaround which I discourage to use ;-) Maybe it makes the problem/solution clearer:

```
public function parseTemplateHook(Template $template)
    {
        // fix preview urls (see https://github.com/contao/contao/issues/2580)
        if ($template->getName() !== 'be_main' || $this->request->getGet('act') !== 'edit' ||
            !($table = $this->request->getGet('table')) || !($id = $this->request->getGet('id')) ||
            !in_array($table, ['tl_news', 'tl_calendar_events'])) {
            return;
        }

        $data = $template->getData();

        if ($table === 'tl_calendar_events') {
            $insertTag = '{{event_url::' . $id . '}}';

            $data['frontendFile'] = '?url=' . urlencode(Controller::replaceInsertTags($insertTag));
        }

        if ($table === 'tl_news') {
            $insertTag = '{{news_url::' . $id . '}}';

            $data['frontendFile'] = '?url=' . urlencode(Controller::replaceInsertTags($insertTag));
        }

        $template->setData($data);
    }
```

**Comment by Defcon0** [comment]
Post to avoid stale :-)

**Comment by SeverinGloeckle** [comment]
I have checked this issue again in Contao 4.9.

While the original issue and fix was based on Contao 4.4 and Contao 4.9 has changed the preview URL generation, the bug still exists:
- Go to Events -> open calendar -> edit event settings
- Start the frontend preview -> correct redirect to the event detail page
- Logout / destroy the backend session
- Reload the event edit view
- Log into the Contao backend again
- Start the frontend preview again -> incorrect redirect to frontpage

The correct URL in the preview is:
````
contao/preview?calendar=1
````

The incorrect URL after re-login is:
````
contao/preview
````


**More debugging:**
- The PreviewUrlCreateEvent is not dispatched in BackendPreviewListener
- This is due to the CURRENT_ID value missing from the session
- The CURRENT_ID should be set in Backend::getBackendModule

````php
$id = (!Input::get('act') && Input::get('id')) ? Input::get('id') : $objSession->get('CURRENT_ID');
````

The CURRENT_ID is not set because we are already in the 'edit view' with `act=edit`, so the CURRENT_ID is loaded from the session and not from the GET parameter.

**This suggests that the problem is not limited to the frontend preview, but affects all code that relies on the CURRENT_ID parameter of the session.**
